### PR TITLE
Change MIDI velocity implementation to allow direct control of velocity value

### PR DESCRIPTION
**Pull request by jakobkg**
## Description

I've been trying to make a MIDI keyboard based on QMK, and one of the challenges of using mechanical keyboard switches for this is the lack of velocity sensitivity. To compensate, I was trying to add velocity control with an analog input, like a potentiometer. To my disappointment, I wasn't able to use my potentiometer to change velocity freely between 0-127, as the current implementation forces velocity down to a range of 1-10 which is far too low resolution for my wants and needs. These changes are an attempt at fixing this without breaking the current velocity keycodes, to allow the old method to work, and open up for setting velocity directly with something as simple as
`midi_config.velocity = analogReadPin(POT_PIN) >> 3`

My changes are at a relatively low level so I wanted to make this PR early so I don't put too much work in only to end up reworking them if they don't mesh well with the ongoing refactoring. As far as I can tell these parts of QMK are barely being used by any keymaps so I'm not super worried about them, but I understand that changes might be needed to preserve existing behaviour even better than what I have currently implemented.

As a side effect of my changes, the function `compute_velocity()` in process_midi.c is now only called when velocity is actually changed via the related keycodes, instead of being called on every MIDI note on or note off being message sent which should help performance a tiny bit.

I'm currently unsure if I have reached parity with how the existing MIDI_VELOCITY_MAX and MIDI_VELOCITY_MIN values are being used, but they don't seem to be very well implemented at the moment as it's possible to repeatedly press MI_VELU and have velocity overflow past 127 and wrap back to 0 which is very jarring. This behavior has been changed in this PR, to limit velocity values to 0-127 when being changed using these keycodes.

## Types of Changes

- [ ] Core
- [X] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* None, this is an expansion on an existing, but not much used, feature

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/process_keycode/process_midi.c [hunk 1]
```diff
@@ -122,19 +122,30 @@ bool process_midi(uint16_t keycode, keyrecord_t *record) {
...
             if (record->event.pressed) {
-                midi_config.velocity = keycode - MIDI_VELOCITY_MIN;
+                midi_config.velocity = compute_velocity(keycode - MIDI_VELOCITY_MIN);
                 dprintf("midi velocity %d\n", midi_config.velocity);
             }
             return false;
         case MI_VELD:
             if (record->event.pressed && midi_config.velocity > 0) {
-                midi_config.velocity--;
+                if (midi_config.velocity > 12 ) {
```
Comment (fauxpark) [inline quantum/process_keycode/process_midi.c:131]:
```suggestion
                if (midi_config.velocity > 12) {
```

File: quantum/process_keycode/process_midi.c [hunk 2]
```diff
@@ -122,19 +122,30 @@ bool process_midi(uint16_t keycode, keyrecord_t *record) {
...
+                    midi_config.velocity -= 13;
+                }
+                else {
```
Comment (fauxpark) [inline quantum/process_keycode/process_midi.c:134]:
```suggestion
                } else {
```

File: quantum/process_keycode/process_midi.c [hunk 3]
```diff
@@ -122,19 +122,30 @@ bool process_midi(uint16_t keycode, keyrecord_t *record) {
...
+                }
+
                 dprintf("midi velocity %d\n", midi_config.velocity);
             }
             return false;
         case MI_VELU:
-            if (record->event.pressed) {
-                midi_config.velocity++;
+            if (record->event.pressed && midi_config.velocity < 127) {
+                if (midi_config.velocity < 115 ) {
```
Comment (fauxpark) [inline quantum/process_keycode/process_midi.c:143]:
```suggestion
                if (midi_config.velocity < 115) {
```

File: quantum/process_keycode/process_midi.c [hunk 4]
```diff
@@ -122,19 +122,30 @@ bool process_midi(uint16_t keycode, keyrecord_t *record) {
...
+                    midi_config.velocity += 13;
+                }
+                else {
```
Comment (fauxpark) [inline quantum/process_keycode/process_midi.c:146]:
```suggestion
                } else {
```

**Comment by jakobkg** [comment]
I've done some further testing and the velocity values you can get using MI_VELU and MI_VELD before/after these changes are similar enough for this not to be disruptive to existing keyboards/keymaps.

With the current implementation, the velocity values available are:
`12, 24, 36, 48, 60, 72, 84, 96, 108, 120`
Continuing to press MI_VELU will allow the value to overflow past 127, which is undesired at best and could cause real problems if the receiving software doesn't handle the out-of-range velocity values gracefully.

With my implementation, the values available using MI_VELU/MI_VELD are:
`0, 13, 26, 39, 52, 65, 78, 91, 104, 117, 127`
Continuing to press MI_VELU has no effect once 127 is reached.

I recognize that this is 11 values and won't map well to the existing keycodes for MI_VEL_[1-10]. I think the solution here is to simply add the keycode MI_VEL_0, which sets MIDI velocity to 0.

**Review by tzarc — APPROVED** [review]

File: quantum/process_keycode/process_midi.h [hunk 5]
```diff
@@ -35,7 +35,7 @@ typedef union {
     struct {
         uint8_t octave : 4;
         int8_t  transpose : 4;
-        uint8_t velocity : 4;
+        uint8_t velocity;
```
Comment (drashna) [inline quantum/process_keycode/process_midi.h:38]: Why the change here?

Comment (fauxpark) [reply quantum/process_keycode/process_midi.h:38]: The range of `velocity` is now 0-127 (as it should be), not 0-15.

Comment (drashna) [reply quantum/process_keycode/process_midi.h:38]: so, this could be `:7`, but simpler to just not.

Comment (jakobkg) [reply quantum/process_keycode/process_midi.h:38]: I originally figured readability was worth more than the single bit of memory that might be saved, but upon further consideration I think there is value to using `:7` here. Especially with how it enables more graceful handling of user-caused overflows. Now that users will be able to control `midi_config.velocity` directly there is nothing stopping them from going past 127 with something like an errant loop or misconfigured rotary encoder. It would probably be nicer of us to have the value overflow back to 0 once it tries to go past 127 than to allow out-of-range velocity values to be set in the first place.

I feel like it's best to protect users and other software and equipment further down the signal chain from invalid MIDI messages, which means either changing this line to `uint8_t velocity : 7;` or implementing some other check in `process_midi()` or `midi_send_noteon()`. Out of these options, `uint8_t velocity : 7;` is probably the simplest one

**Review by noroadsleft — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!
